api/v1beta1/bucket_types.go

```diff
@@ -30,6 +28,18 @@ const (
 	BucketKind = "Bucket"
 )
 
+const (
+	GenericBucketProvider string = "generic"
+	AmazonBucketProvider  string = "aws"
+)
+
+const (
+	// DownloadFailedCondition indicates a transient or persistent download failure. If True, observations on the
+	// upstream Source revision are not possible, and the Artifact available for the Source may be outdated.
+	// This is a "negative polarity" or "abnormal-true" type, and is only present on the resource if it is True.
+	DownloadFailedCondition string = "DownloadFailed"
```

Comment:
To simplify the API and ease observability, I think this condition could be consolidated with `CheckoutFailed` to a more generic Condition type which can be used for any Source which wants to reflect a failure during the connection to, or retrieval from the upstream source.

The question is what to name it, and if it should actually be related to the specific generic action ("retrieving" something, but this may be an assumption?), or more about the fact that (part) of the source is just not available.

In case of the latter, the Reason and Message (or additional context from an Event) could then provide further information about the issue (e.g. a networking error) to a consumer. In this case, the Type could simply be `UpstreamUnavailable` (or `UpstreamAvailable`).

Another question this raises is if there is a need to distinguish "unavailable due to an invalid reference" from "unavailable due to a network error". Note hereby, that while Reasons do provide context to humans, they are not considered part of the API.

Reply:
`FetchFailed` perhaps. I think it _is_ worth distinguishing problems that are probably due to the server or network conditions, from problems that may indicate incorrect user input.

Reply:
I think `FetchFailed` is generic enough and signals the underlying issue 👍